Mage.Sets/src/mage/cards/g/GuffRewritesHistory.java

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+}
+
+class GuffRewritesHistoryEffect extends OneShotEffect {
+
+    public GuffRewritesHistoryEffect() {
+        super(Outcome.Neutral);
+    }
+
+    private GuffRewritesHistoryEffect(final GuffRewritesHistoryEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public GuffRewritesHistoryEffect copy() {
+        return new GuffRewritesHistoryEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        // set of controllers for the second part of the effect.
+        Set<UUID> controllers = new HashSet<>();
+        Map<UUID, Cards> toShuffle = new HashMap<>();
+        for (Target target : source.getTargets()) {
+            for (UUID permId : target.getTargets()) {
+                Permanent permanent = game.getPermanent(permId);
+                if (permanent == null) {
+                    continue;
+                }
+                Player owner = game.getPlayer(permanent.getOwnerId());
+                Player controller = game.getPlayer(permanent.getControllerId());
+
+                if (controller != null) {
+                    controllers.add(controller.getId());
+                }
+
+                if (owner != null) {
+                    toShuffle.computeIfAbsent(owner.getId(), k -> new CardsImpl());
+                    toShuffle.get(owner.getId())
+                             .add(permanent);
+                }
+            }
+        }
+
+        // Only one shuffle per owner!
+        for (UUID playerId : game.getState().getPlayersInRange(source.getControllerId(), game)) {
+            if (toShuffle.containsKey(playerId)) {
+                Player owner = game.getPlayer(playerId);
+                // Those permanents' owners shuffle them into their libraries.
+                owner.shuffleCardsToLibrary(toShuffle.get(playerId), game, source);
+            }
+        }
+
+        // Nonland card for each player for the third part of the ability.
+        Map<UUID, Card> nonlands = new HashMap();
+
+        for (UUID playerId : game.getState().getPlayersInRange(source.getControllerId(), game)) {
+            if (!controllers.contains(playerId)) {
+                continue;
+            }
+
+            Player controller = game.getPlayer(playerId);
+            if (controller == null) {
+                continue;
+            }
+
+            Cards cards = new CardsImpl();
+            Card nonland = null;
+            for (Card card : controller.getLibrary().getCards(game)) {
+                cards.add(card);
+                if (!card.isLand(game)) {
+                    nonland = card;
+                    break;
+                }
+            }
+
+            // Each player who controlled one of those permanents exiles cards from the top of their library
+            // until they exile a nonland card, then puts the rest on the bottom of their library in a random order.
+            controller.moveCards(cards, Zone.EXILED, source, game);
```

Comment:
may need some `retainZone(Zone.EXILED, game);` ?
Still not understanding fully how the replacement effects on card moving zone are working.

Reply:
As I understand it, `retainZone()` is just for excluding cards that are no longer in that zone (e.g. because the effect moved one of them elsewhere). I think you've handled it already when you call `cards.remove(nonland)`

Reply:
Ah I see now, you're concerned that the cards might not actually be in the exiled zone due to some replacement effect? Is there an example? If so then yeah I guess the retainZone check would be needed to catch that edge case.

Reply:
I guess there is no effect right now, but a card with "If you would exile {this} from your library, put it into play instead" would break things.

I'll add the `retainZone()` between the exile and the shuffling back, at worst it does nothing but waste a tiny amount of CPU time.